docs/src/ref/modeling.md

```diff
@@ -97,31 +97,46 @@ val::Bool = @trace(bernoulli(0.5), :z)
-For example, consider the following generative function:
+
+### Sample space and support of random choices
+
+Different probability distributions produce different types of values for their random choices. For example, the [`Bernoulli`](@ref) distribution results in `Bool` values (either `true` or `false`), the [`Normal`](@ref) distribution results in `Real` values that may be positive or negative, and the [`Beta`](@ref) distributions result in `Real` values that are always in the unit interval (0, 1).
+
+Each [`Distribution`](@ref) is associated with two sets of values:
+
+- The **sample space** of the distribution, which does not depend on the arguments.
+
+- The **support** of the distribution, which may depend on the arguments, and is the set of values that has nonzero probability (or probability density). It may be the entire sample space, or it may be a subset of the sample space.
+
+For example, the sample space of [`Bernoulli`](@ref) is `Bool` and its support is either `{true}`, `{false}`, or `{true, false}`. The sample space of [`Normal`](@ref) is `Real` and its support is the set of all values on the real line. The sample space of [`Beta`](@ref) is `Real` and its support is the set of values in the interval (0, 1).
+
+Gen's built in modeling languages require that a address is associated with a fixed sample space. For example, it is not permitted to use a `Bernoulli` distribution to sample at addresss `:a` in one execution, and a `Normal` distribution to sample at address `:a` in a different execution, because their sample spaces differ (`Bool` vs `Real`):
 ```julia
 @gen function foo()
-    n = @trace(categorical([0.5, 0.5]), :n) + 1
-    @trace(categorical(ones(n) / n), :x)
+    if @trace(bernoulli(0.5), :branch)
+        @trace(bernoulli(0.5), :x)
+    else
+        @trace(normal(0, 1), :x)
+    end
 end
 ```
-The support of the random choice with address `:x` is either the set ``\{1, 2\}`` or ``\{1, 2, 3\}``.
-Therefore, this random choice does satisfy our condition above.
-This would cause an error with the following, in which the `:n` address is modified, which could result in a change to the domain of the `:x` variable:
-```julia
-tr, _ = generate(foo, (), choicemap((:n, 2), (:x, 3)))
-tr, _ = mh(tr, select(:n))
-```
-We can modify the address to satisfy the condition by including the domain in the address:
+
+A generative function can be **disciplined** or not. In a disciplined generative function, the support of random choices at each address must be fixed. That is, there exists a set S that is a subset of the sample space such that for all executions of the generative function, if an address appears in the execution, its support must be S. Violating this discipline will cause NaNs, errors, or undefined behavior in certain inference programs. However, there are many inference programs where violating this discipline does not result in undefined behavior and is convenient.
```

Comment:
```suggestion
A generative function can be **disciplined** or not. In a disciplined generative function, the support of random choices at each address must be fixed. That is, for each address `a` there exists a set S that is a subset of the sample space such that for all executions of the generative function, if `a` occurs as the address of a choice in the execution, the support of that choice must be exactly S. Violating this discipline will cause NaNs, errors, or undefined behavior in some inference programs. However, there are many inference programs where violating this discipline does not result in undefined behavior and is convenient.
Authors of inference programs should document which kinds of undisciplined models their inference algorithms allow or expect to see.
```

Reply:
I don't think we should prescribe what authors of inference programs should do, since inference programs are user code, and are not necessarily documented at all.

Reply:
We could qualify this with "authors of reusable inference code" ..

Reply:
Sure.
```suggestion
A generative function can be **disciplined** or not. In a disciplined generative function, the support of random choices at each address must be fixed. That is, for each address `a` there exists a set S that is a subset of the sample space such that for all executions of the generative function, if `a` occurs as the address of a choice in the execution, the support of that choice must be exactly S. Violating this discipline will cause NaNs, errors, or undefined behavior in some inference programs. However, there are many inference programs where violating this discipline does not result in undefined behavior and is convenient.
Authors who want their inference code to be reusable should consider documenting which kinds of undisciplined models their inference algorithms allow or expect to see.
```